### PR TITLE
[FLINK-5720] Deprecate DataStream#fold()

### PR DESCRIPTION
**Pull request by zentol**
This PR deprecates the various `DataStream#fold()` methods and all related internal classes.

### CODE REVIEW

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala [hunk 1]
```diff
@@ -183,7 +183,10 @@ class KeyedStream[T, K](javaStream: KeyedJavaStream[T, K]) extends DataStream[T]
    * Creates a new [[DataStream]] by folding the elements of this DataStream
    * using an associative fold function and an initial value. An independent 
    * aggregate is kept per key.
+   *
+   * @deprecated will be removed in a future version
    */
+  @Deprecated
```
Comment (aljoscha) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: Scala has `@deprecated` where you can also specify the reason right in the annotation. (https://coderwall.com/p/k4ulyg/scala-deprecated-annotation)

Also holds for the other instances in Scala code.

Comment (zentol) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: Do you propose to only have the reason in the annotation, or in both annotation and docs?

**Comment by zentol** [comment]
Man that stuff is everywhere isn't it...

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala (hunk 1 above)
Comment (aljoscha) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: I don't know, actually. It seems actual (random) code on GitHub only has it in the annotation: https://github.com/ReactiveX/RxScala/blob/0.x/src/main/scala/rx/lang/scala/Observable.scala#L4876

Comment (zentol) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/KeyedStream.scala]: I've included them in the annotation.

**Comment by aljoscha** [comment]
I think we got them all now. I would have missed the one in `TypeExtractor`...

LGTM (looks good to merge) now!

**Comment by zentol** [comment]
merging.

**Comment by aljoscha** [comment]
Cool, thanks for working on this (and the fold PR): 👍

**Comment by caddac** [comment]
Sorry for commenting on this closed PR, but can you point me to documentation for what will replace fold? I'm starting a new project using 1.3 and don't want to use a deprecated API, but don't find much on what I should be replacing fold with.  Thanks

**Comment by wuchong** [comment]
Hi @caddac ,  you can use `AggregateFunction` and `KeyedStream.aggregate(AggregationFunction<T> aggregate)` instead of fold.

**Comment by NicoK** [comment]
Actually, `KeyedStream.aggregate` is protected and thus not accessible.

You may, however, access `WindowedStream#aggregate(AggregateFunction<T,ACC,R>) ` and `AllWindowedStream#aggregate(AggregateFunction<T,ACC,R>) ` for windowed streams and use a stateful map or flatmap for the rest.

**Comment by iduanyingjie** [comment]
In this document
[https://ci.apache.org/projects/flink/flink-docs-release-1.4/dev/stream/operators/index.html#datastream-transformations](url)
`keyedStream.fold("start")((str, i) => { str + "-" + i })`
I want to do the same thing, what am I supposed to do and and don't use deprecated API

**Comment by fhueske** [comment]
Hi @iduanyingjie, the replacement for fold has not been implemented yet. So you have to stick with `fold()` for now.

`fold`'s replacement will correspond to `WindowedStream.aggregate(AggregateFunction)`.

**Pull request closed without merging**